### PR TITLE
TypeScript conversion

### PR DESCRIPTION
**Pull request by StrahilKazlachev**
- [ ] - review API changes
- [ ] - review typings
- [ ] - review build and test setup - check what more is needed for the `doc` task
- [x] - finish tests conversion
- [x] - fill in missing API doc
- [x] - ~~convert sample - src and setup~~
- [ ] - update *how to...* in *readme*

### CODE REVIEW

**Comment by StrahilKazlachev** [comment]
- Got the build setup from `aurelia-validation`.
- main API change
  - `DialogService` no longer adds undeclared properties to the `DialogController` instances - `view: View`, `viewModel: AnyUserProvidedViewModel`, `slot: ViewSlot`
  - added `controller: Controller` to the `DialogController` API - the `DialogController` uses it to call `.unbind`, the `Renderer`s to call `.attached` and `.dettached`.
- typings issue - I'm having hard time properly typing `DialogService.open`. The settings `yieldController` and `rejectOnCancel` do modify the output of the method, but when set in the plugin configuration I can't think of a way to tell TS what the output should be.

File: package.json [hunk 1]
```diff
@@ -15,11 +15,41 @@
...
+    "postbuild:es2015": "cross-env copyfiles --up 1 src/**/*.html src/**/*.css dist/es2015",
+    "prebuild:native-modules": "cross-env rimraf dist/native-modules",
+    "build:native-modules": "cross-env tsc --project tsconfig.build.json --outDir dist/native-modules --module es2015",
+    "postbuild:native-modules": "cross-env copyfiles --up 1 src/**/*.html src/**/*.css dist/native-modules",
+    "prebuild:system": "cross-env rimraf dist/system",
+    "build:system": "cross-env tsc --project tsconfig.build.json --outDir dist/system --module system",
+    "postbuild:system": "cross-env copyfiles --up 1 src/**/*.html src/**/*.css dist/system",
+    "prebuild": "cross-env rimraf dist",
+    "build": "concurrently \"npm run build:amd\" \"npm run build:commonjs\" \"npm run build:es2015\" \"npm run build:native-modules\" \"npm run build:system\"",
+    "predoc": "cross-env rimraf doc/api.json && rimraf dist/doc-temp && tsc --project tsconfig.build.json --outFile dist/doc-temp/aurelia-validation.js && node doc/shape-defs && copyfiles tsconfig.json dist/doc-temp",
```
Comment (PWKad) [inline package.json]: `aureliva-validation` should be `aurelia-dialog` here I think

File: package.json [hunk 2]
```diff
@@ -15,11 +15,41 @@
...
+    "doc": "cross-env typedoc --json doc/api.json --excludeExternals --includeDeclarations --mode modules --target ES6 --name aurelia-validation-docs dist/doc-temp/",
```
Comment (PWKad) [inline package.json]: Same here.

File: src/ai-dialog-footer.ts [hunk 3]
```diff
@@ -1,44 +1,46 @@
+import {inject} from 'aurelia-dependency-injection';
```
Comment (PWKad) [inline src/ai-dialog-footer.ts]: I typically favor `static inject = []` over the decorator, not sure if there is any real speed implication but for less imports but that's more of a style choice just pointing it out.

Comment (PWKad) [inline src/ai-dialog-footer.ts]: Also one note is you can put it right beside the constructor which improves legibility and refactoring.

File: src/ai-dialog-footer.ts [hunk 4]
```diff
@@ -1,44 +1,46 @@
...
+ * View-model for footer of Dialog.
+ */
 @customElement('ai-dialog-footer')
+/* tslint:disable:max-line-length */
 @inlineView(`
   <template>
     <slot></slot>
-
     <template if.bind="buttons.length > 0">
       <button type="button" class="btn btn-default" repeat.for="button of buttons" click.trigger="close(button)">\${button}</button>
```
Comment (PWKad) [inline src/ai-dialog-footer.ts]: We can make this - 

```
<button type="button" class="btn btn-default"
  repeat.for="button of buttons"
  click.trigger="close(button)">\${button}</button>
```
To comply to the rule.  I'd rather not have files disabling linting in them unless we absolutely need to (such as copying over some open source file and not wanting to modify the source)

File: src/ai-dialog-footer.ts [hunk 5]
```diff
@@ -1,44 +1,46 @@
...
+@inject(DialogController)
 export class AiDialogFooter {
-  static inject = [DialogController];
+  public static isCancelButton(value: string) {
+    return value === 'Cancel';
+  }
 
-  @bindable buttons: any[] = [];
-  @bindable useDefaultButtons: boolean = false;
+  @bindable()
```
Comment (PWKad) [inline src/ai-dialog-footer.ts]: `@bindable public buttons: any[] = [];`

Should work here, was there an issue with the decorator here?

File: src/aurelia-dialog.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import {FrameworkConfiguration} from 'aurelia-framework';
+import {DialogConfiguration} from './dialog-configuration';
+
+ // tslint:disable-next-line:max-line-length
+export function configure(frameworkConfig: FrameworkConfiguration, callback?: (config: DialogConfiguration) => void): void {
```
Comment (PWKad) [inline src/aurelia-dialog.ts]: Again here I think -

```
export function configure(
  frameworkConfig: FrameworkConfiguration, 
  callback?: (config: DialogConfiguration) => void): void {
```

is favorable

File: src/dialog-configuration.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
...dth:fit-content;height:-moz-fit-content;height:-webkit-fit-content;height:fit-content;margin:auto;border-image-source:initial;border-image-slice:initial;border-image-width:initial;border-image-outset:initial;border-image-repeat:initial;background:#fff}ai-dialog>ai-dialog-header{display:block;padding:16px;border-bottom:1px solid #e5e5e5}ai-dialog>ai-dialog-header>button{float:right;border:none;display:block;width:32px;height:32px;background:0 0;font-size:22px;line-height:16px;margin:-14px -16px 0 0;padding:0;cursor:pointer}ai-dialog>ai-dialog-body{display:block;padding:16px}ai-dialog>ai-dialog-footer{display:block;padding:6px;border-top:1px solid #e5e5e5;text-align:right}ai-dialog>ai-dialog-footer button{color:#333;background-color:#fff;padding:6px 12px;font-size:14px;text-align:center;white-space:nowrap;vertical-align:middle;-ms-touch-action:manipulation;touch-action:manipulation;cursor:pointer;background-image:none;border:1px solid #ccc;border-radius:4px;margin:5px 0 5px 5px}ai-dialog>ai-dialog-footer button:disabled{cursor:default;opacity:.45}ai-dialog>ai-dialog-footer button:hover:enabled{color:#333;background-color:#e6e6e6;border-color:#adadad}.ai-dialog-open{overflow:hidden}`;
```
Comment (PWKad) [inline src/dialog-configuration.ts]: Seems reasonable here :)

File: src/dialog-renderer.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+const containerTagName = 'ai-dialog-container';
+const overlayTagName = 'ai-dialog-overlay';
+
+export const transitionEvent = (() => {
+  let transition: string | undefined;
+  return (): string => {
+    if (transition) { return transition; }
+    const el = DOM.createElement('fakeelement') as HTMLElement;
+    /* tslint:disable:object-literal-key-quotes */
```
Comment (PWKad) [inline src/dialog-renderer.ts]: Any reason to just convert this to not use the string keys?

**Comment by PWKad** [comment]
Left some initial feedback (more style related) going to pull down now and start looking at everything else nice job though!

**Comment by jods4** [comment]
Great to see more of aurelia moving to Typescript! 
It makes it much easier to work on the codebase and produces good `.d.ts`.

You may want to integrate with `aurelia-tools`? There are some base configs and tasks over there, already used by `aurelia-loader-webpack`.

One critical change is that you absolutely need to depend on external TS helpers (`tslib`) so that they are not emitted in every file. But there was some discussion over that, see issue aurelia/tools#52.

**Comment by StrahilKazlachev** [comment]
@PWKad So I was not happy how the `yieldController` setting changed the flow of `.open`. So I defined new `DialogOpenPromise` - essentially `Promise<T> & {whenClosed: Function}`(check *src/dialog-service.ts*), that gets returned by the `.open` method. So now `.open().then()` resolves when the dialog is open and gives you the `DialogOpenResult`, and `open().whenClosed()` is a shortcut for something like `.open().then(r => r.closeResult).then()`. Also `.whenClosed` has the signature of `Promise<DialogCloseResult>.then`.

**Comment by PWKad** [comment]
@StrahilKazlachev sounds good do you need me to jump in on any of the tasks in the first comment?

**Comment by StrahilKazlachev** [comment]
@PWKad 
- I can't do the API and typings review.
- The update to the *readme* should probably be done after your PR is in, as the related updates to *doc/article* - API changes and examples.
- not sure what to do with the sample - how to set it up in the repo, or to switch to runnable gist, ... Any ideas?
- about the doc task it seems I just have to copy some files from the validation repo  - `shape-doc.js` and `shape-defs.js` - @niieani is that correct, also that should do till you manage to finish aurelia-tools, right?

**Comment by StrahilKazlachev** [comment]
@PWKad Almost forgot, could you also review the changes to *circle.yml* if you haven't.

**Comment by niieani** [comment]
@StrahilKazlachev I think you could use `aurelia-tools` already (installable from Git, not the NPM published package), as it contains the shape tools. The remaining tasks for aurelia-tools are pretty minor as things there mostly work as they should.

**Comment by StrahilKazlachev** [comment]
@niieani I tried `aurelia-tools` and the `doc` task didn't work. I got the same error when I tried to run it on `loader-webpack`.
```
> aurelia-tools doc-build --continue-when-failed

events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: spawn ./node_modules/.bin/tsc ENOENT
    at exports._errnoException (util.js:1026:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:359:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
    at Module.runMain (module.js:592:11)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

**Comment by niieani** [comment]
Do you have typescript installed as a devDependency? 
The error seems to be that `./node_modules/.bin/tsc` is missing. Can you verify if the file exists?
Also, what `pwd` are you running the command from? And which version of aurelia-tools is the dependency?

**Comment by PWKad** [comment]
@StrahilKazlachev let's remove the sample for now.  I'll open a new issue about getting the gist.run up and running with the dialog for repro's.

**Comment by PWKad** [comment]
Also any ideas how to run the tests now?  I'm getting failures for everything where it seems like the components are not being created to test.  Let me know and I'll try to help pick up on the testing side.

**Comment by StrahilKazlachev** [comment]
Hmm, all tests pass for me. I just run `npm run test` or `npm run test-watch`. Could you try `npm run test` if you haven't? Also I'll be redoing the tasks to use `aurelia-tools` tonight or tomorrow.

**Pull request merged**